### DISCUSSION TITLE
Need an advice how to extend ltree support for subqueries

### DISCUSSION THREAD

**Discussion by mlesin** [diesel-rs/diesel#3394] [Q&A]
I'm using [diesel_ltree](https://github.com/kivikakk/diesel_ltree) crate for adding support of ltree into diesel, and facing an issue with it.
Currently, I can't compile query like this:
```sql
SELECT *
FROM my_tree mt
WHERE  mt."path" @> ANY
    (
        SELECT mt2."path"
        FROM my_tree mt2
        WHERE mt2."path" <@ 'root.archaea.thermoprotei' 
    )
```

I'm defining subquery separately, and then trying to add it into main query with `.filter(my_tree::path.contains_any(subquery))` but it doesn't compile, while, at the same time, if I change subquery to return id's and change filter to something like `.filter(my_tree::id.eq_any(subquery))` it compiles fine.

I thought to modify diesel_ltree crate to support this, so I looked into diesel sources to find how it is implemented for `eq_any`, but found it is accomplished with `AsInExpression` trait, but it seems to be in private area of diesel and therefore not available for using in external diesel_ltree crate.
**EDIT:** I found `AsInExpression` trait can be unlocked with `i-implement-a-third-party-backend-and-opt-into-breaking-changes` feature, but there is other useful trait `Grouping` which is needed too, but it's still in private area even with this feature enabled.

I'm not sure I'm digging into right direction at all, so any advice would help a lot.

(I opened [issue](https://github.com/kivikakk/diesel_ltree/issues/22) in diesel_ltree crate as well with more details about a problem)

**EDIT:** By the way, the more I dig into this, the more it seems to me that it is more correct to implement Ltree support in diesel itself, rather than in a separate crate because of too tight connections to the inner structure of diesel. Am I right?

**Comment by mlesin** [comment]
Also I thinking about how to workaround this problem, maybe there is some way to include subquery's generated SQL with bindings into main query using `sql()` function? But I can't find any way how to do this right.

**Reply by mlesin** [reply]
Digging further, I found in my case the more correct solution (and, regarding to EXPLAIN ANALYZE, more performant) is to use `ARRAY` keyword instead of `ANY`, but diesel's array expression in Pg uses syntax with square brackets instead of round ones, and requires `AsExpressionList` trait, but in this case something like `AsInExpression` is needed. Do I need to create another expression for this case?

**Comment by weiznich** [comment]
> I'm defining subquery separately, and then trying to add it into main query with .filter(my_tree::path.contains_any(subquery)) but it doesn't compile, while,

Diesel does not provide a `contains_any` method at all, so I'm not sure to which method you are referring. Maybe that method is coming from the `diesel_ltree` crate, which would indicate that you should open an issue there?

> EDIT: I found AsInExpression trait can be unlocked with i-implement-a-third-party-backend-and-opt-into-breaking-changes feature, but there is other useful trait Grouping which is needed too, but it's still in private area even with this feature enabled.
>
> I'm not sure I'm digging into right direction at all, so any advice would help a lot.

Just to be very clear: We do not consider traits/types/… behind this feature flag to be part of our stable API. Those are parts of the internal implementation of diesel. They can (and will) change at any future release, so expect breakage if you depend on them. That written: Adding another type/trait/… to the semi-public API behind this feature flag is something that's possible without much effort. I would expect a PR for such changes that outlines a use-case for the added types.

> EDIT: By the way, the more I dig into this, the more it seems to me that it is more correct to implement Ltree support in diesel itself, rather than in a separate crate because of too tight connections to the inner structure of diesel. Am I right?

`Ltree` is not a basic postgres type. Our current rules for these types are that they need to be implemented in third party crates as we just do not have the capacity to maintain support for all possible extensions of postgresql.

**Reply by mlesin** [reply]
> Diesel does not provide a contains_any method at all, so I'm not sure to which method you are referring. Maybe that method is coming from the diesel_ltree crate, which would indicate that you should open an issue there?

Yes, it's from this crate, I opened issue there, and even managed how to implement this change (testing it right now) actual change is [here](https://github.com/mlesin/diesel_ltree/commit/e09eab0f487c02f97f6d9ec2db81c239cb082a63) if you interested in details.
However, the dependency from `AsInExpression` seems to be inevitable in this case, but it is only possible with "this-big-breaking-change-warning" feature. Is there a strong reason to keep this concrete trait hidden from users? It's quite useful and pretty hard to reimplement using only public API to diesel.

> Ltree is not a basic postgres type. Our current rules for these types are that they need to be implemented in third party crates as we just do not have the capacity to maintain support for all possible extensions of postgresql.

So keeping `diesel_ltree` separate is correct. That's good. Maybe we can make some of internal traits visible through public API for this use-case then? I'm interested at least in `AsInExpression`.

**Reply by weiznich** [reply]
Generally speaking: We try to be conservative with what's part of our public API and what's not as any failure there requires a major version bump to fix it. Adding internal AST nodes like `AsInExpression` to the stable public API can be possible, but requires quite a bit of work. We must be sure that we don't want to change these types later. I'm not sure if I personally want to invest the time to evaluate that for `AsInExpression`. If someone else can reasonable describe why this should be public and what are potential problems with having this public we might talk about such "stabilization".

**Reply by mlesin** [reply]
In this exact case I need an ability for a custom type (Ltree), sitting in standalone crate, define an operator @>, right side of which can be either `Ltree` or `ARRAY of Ltree`.
This works only in basic case when array is usual constants, but fails when right part contains subquery wrapped with type aliasing like `ARRAY( SELECT ltree_field FROM table)`
Only solution I found to solve this was using `AsInExpression` trait, but may be you can suggest anything better?

**Reply by weiznich** [reply]
It is always possible to have another trait in `diesel_ltree` for this specific case. This would then be implemented for the accepted types.

**Reply by mlesin** [reply]
Could you please suggest me what trait I should implement (or redeclare) in `diesel_ltree` to solve this problem with subquery without relying on `i-implement-a-third-party-backend-and-opt-into-breaking-changes` feature?

**Reply by weiznich** [reply]
It should be possible to just redeclare `AsInExpression` as well as `Grouped` outside of diesel.
